itou/www/signup/forms.py

```diff
@@ -91,50 +96,164 @@ def save(self, request):
         return user
 
 
-class SiaeSignupForm(FullnameFormMixin, SignupForm):
+class SelectSiaeForm(forms.ModelForm):
+    """
+    First of two forms of siae signup process.
+    This first form allows the user to select which siae will be joined.
+    """
+
+    class Meta:
+        model = Siae
+        fields = ["kind", "siret", "email"]
+
+    kind = forms.ChoiceField(
+        label=_("Type de votre SIAE"),
+        choices=BLANK_CHOICE + Siae.KIND_CHOICES,
+        required=True,
+    )
 
     siret = forms.CharField(
         label=_("Numéro SIRET de votre SIAE"),
+        min_length=14,
         max_length=14,
         validators=[validate_siret],
-        required=True,
         strip=True,
-        help_text=_("Saisissez 14 chiffres."),
+        help_text=_(
+            "Saisissez 14 chiffres. Doit si possible être le SIRET connu de l'ASP."
+        ),
+        required=False,
     )
 
-    def clean_siret(self):
-        siret = self.cleaned_data["siret"]
+    email = forms.EmailField(
+        label=_("E-mail"),
+        help_text=_("Doit si possible être l'adresse e-mail connue de l'ASP."),
+        required=False,
+    )
 
-        siret_exists_in_db = Siae.active_objects.filter(siret=siret).exists()
-        if siret_exists_in_db:
-            return siret
+    def save(self, request, commit=True):
+        raise RuntimeError("SelectSiaeForm.save() should never be called.")
+
+    # pylint: disable=inconsistent-return-statements
+    def clean(self):
+        kind = self.cleaned_data["kind"]
+        siret = self.cleaned_data["siret"]
+        email = self.cleaned_data["email"]
+
+        if not (siret or email):
+            error_message = _(
+                "Veuillez saisir soit un email connu de l'ASP soit un SIRET connu "
+                "de l'ASP."
+            )
+            self.raise_validation_error(error_message, add_suffix=False)
+
+        siaes = Siae.active_objects.filter(kind=kind)
+        if siret and email:
+            # We match siaes having any of the two correct fields.
+            siaes = siaes.filter(Q(siret=siret) | Q(auth_email=email))
```

Comment:
Ici, uniquement 3 cas possibles :

1. `.filter(kind=kind, auth_email=email)`
1. `.filter(kind=kind, siret=siret)`
1. `.filter(kind=kind, siret=siret, auth_email=email)`

Je ne comprends pas l'utilité du `OR` ? Plusieurs entrées du même `kind` peuvent exister pour le même `email`. Si tu renseignes en plus du `kind` un `siret` et un `email`, il ne peut y avoir qu'un résultat.

Le seul cas possible qui devrait pouvoir donner plusieurs résultats dans cette configuration est : `.filter(kind=kind, auth_email=email)`

Correct?

Voici un code avec moins de lignes et des messages d'erreurs simplifiés :

```python
# […]

if not (siret or email):
    error_message = _(
        "Précisez votre email ASP ou votre SIRET ASP."
    )
    raise forms.ValidationError(error_message)

siaes = Siae.active_objects.filter(kind=kind)
if siret:
    siaes = siaes.filter(siret=siret)
if email:
    siaes = siaes.filter(auth_email=email)

# Pas besoin du `.all()`
siaes = list(siaes)

if not siaes:
    error_message = _(
        "Aucun résultat, vérifiez les informations saisies."
    )
    raise forms.ValidationError(error_message)

elif len(siaes) > 1:
    # Comme le couple `siret` et `kind` est unique la
    # seule solution pour avoir plusieurs résultats
    # est d'avoir saisi un email.
    error_message = _(
        "Plusieurs structures partagent cet email, veuillez préciser le SIRET ASP."
    )
    raise forms.ValidationError(error_message)

# Ici tu es certain qu'il ne t'en reste qu'une, donc tu peux prendre la première.
# Tu ne retournes pas la `pk` mais l'objet lui même de façon
# à ne pas avoir besoin à faire une nouvelle requête dans la view.
# `self.selected_siae` est défini à `None` dans `__init__()`
# pour être propre et faire plaisir à pylint.
self.selected_siae = siaes[0]
```

La méthode `clean` ne doit rien faire de plus. Le `pylint: disable=inconsistent-return-statements` aurait dû te mettre la puce à l'oreille en terme de *code smell*. Il te manque aussi un [`cleaned_data = super().clean()`](https://docs.djangoproject.com/en/2.2/ref/forms/validation/#validating-fields-with-clean) pour ne pas zapper la validation par défaut.

Si on voulait vraiment simplifier le process pour un utilisateur, alors il faudrait le faire en deux temps :

- d'abord permettre de saisir uniquement un email
- puis proposer en retour une liste de choix possible le cas échéant

Le "Contactez-nous si vous rencontrez des problèmes pour vous inscrire" devrait apparaître de façon permanente sous le formulaire par exemple, les variables pour construire l'email sont passées à tous les templates.

Reply:
Je vois où est le malentendu.

Je pars du principe que l'utilisateur va galérer pour identifier sa SIAE et qu'on doit donc lui donner le maximum de chances de trouver sa SIAE avec le moins d'essais possibles, sachant qu'il ne sait pas si son email est le bon ni si son siret est le bon. C'est pour cela que dans le cas où il entre un email *et* un SIRET (qui était le cas principal et unique avant que j'implémente la possibilité de remplir seulement un des deux champs sur ta suggestion) on matche toutes les SIAE qui ont ce SIRET *ou* cet email.

C'est aussi ce que dit le diagramme cf
![image](https://user-images.githubusercontent.com/10533583/74745243-3eb1b280-526c-11ea-8578-cccb4a6b1158.png)
c'est un OR et non un AND.

Un AND serait beaucoup trop restrictif à mon avis et n'aurait aucune chance de matcher une SIAE en pratique vu que ni nos SIRET ni nos emails ne sont de bonne qualité.

Ceci étant dit, je vais appliquer le reste des recos de ton dernier message.

Reply:
Pour mémoire cela ne se passe pas bien quand je tente un `super().clean()` je me prends des `Un object Structure d&#39;ins'
 b'ertion par l&#39;activit\xc3\xa9 \xc3\xa9conomique avec ces champs Siret et '
 b'Type existe d\xc3\xa9j\xc3\xa0.` et en fait c'est logique car cette form de sélection est pas tout à fait réglo dans le sens où elle n'a pas vocation à créer un nouvel objet siae mais juste à en identifier un existant.

Du coup `super().clean()` cherche à garantir l'unicité (kind, siret) ce que je cherche précisément à contourner puisque je ne suis pas en train de créer une nouvelle siae, mais de vérifier si une telle siae existe.

Reply:
J'ai appliqué toutes tes autres recommandations, sauf le wording que je garde plutôt pour JB.

Reply:
T'aurais pu le faire car ton wording est vraiment à iech ! 😛